### PR TITLE
log viewer and tailer enhancement

### LINKED ISSUE #1588: Configurable job log viewer command per batch system per host
For example, PBS is configured to use a holding area for writing job out/err files while the job is running, and only copy the out/err files to the specified location when the job completes. Local system administrators have developed commands for users to retrieve and print the contents of the out/err files while the job is running. (E.g. `qcat -o|-e JOB-ID`.)

It should be possible for `cylc cat-log` and `cylc gui` to take advantage of this kind of commands, but we'll need to be able to site global configure the command for job hosts with the relevant batch systems.

### PR DESCRIPTION
**Pull request by matthewrmshin**
* cylc cat-log: can now be configured to view `job.out` and `job.err` using
  special commands while the job is running on PBS.
* cylc gui: can now tail follow remote job log files.
* cylc gui: can now be configured to tail follow `job.out` and `job.err`
  using special commands while the job is running on PBS.

Proposed `global.cfg`:

```cfg
[hosts]
    [[myhpc*]]
        [[[batch systems]]]
            [[[[pbs]]]]
                err tailer = qcat -f -e %(job_id)s
                out tailer = qcat -f -o %(job_id)s
                err viewer = qcat -e %(job_id)s
                out viewer = qcat -o %(job_id)s

```

Close #1588.

### CODE REVIEW

**Comment by matthewrmshin** [comment]
@cylc/core This is not yet ready, as it lacks documentation and tests, but some initial comments on the proposed configuration and some quick tests of the GUI tail-follow functionality will be appreciated.

**Comment by hjoliver** [comment]
Looks good.  Are you going to add tailer command templates for other batch systems, or is further investigation required to find out if it is needed or supported in each case?

**Comment by matthewrmshin** [comment]
> Looks good. Are you going to add tailer command templates for other batch systems, or is further investigation required to find out if it is needed or supported in each case?

@hjoliver I'll probably use a `__MANY__` in place of PBS. (As far as we know, `loadleveler` and `slurm` don't need this. I have not looked at `lsf`, but I suspect that it is OK as well. That leaves `moab` and `sge`.)

File: bin/cylc-cat-log [hunk 1]
```diff
@@ -29,227 +29,271 @@ if remrun().execute():
...
-        elif options.stderr:
-            fpath = os.path.join(log_dir, "err")
+    if is_remote_host(host) or is_remote_user(owner):
+        if host and owner:
+            user_at_host = owner + "@" + host
+        elif host:
+            user_at_host = host
+        elif owner:
+            user_at_host = owner + "@localhost"
+    if (options.list_mode or
```
Comment (arjclark) [inline bin/cylc-cat-log]: This looks horrendous...

Comment (matthewrmshin) [inline bin/cylc-cat-log]: I want to continue only if I am not in list or location mode, and the file is "job.out" or "job.err" while the job is running, and the batch system and the batch system job ID are known. Yes, it is complicated.

**Comment by arjclark** [comment]
Nominally looks ok to me - though as discussed some of this looks somewhat complex. Will await the documentation and tests.

**Comment by matthewrmshin** [comment]
Docs and tests added. Branch re-based against current master.

**Comment by matthewrmshin** [comment]
@hjoliver @arjclark please review.

File: bin/cylc-cat-log [hunk 2]
```diff
@@ -29,227 +29,276 @@ if remrun().execute():
...
+              " (file present after an edit-run)."),
+        action="store_const", const=NAME_JOB_EDIT_DIFF, dest="filename")
 
     parser.add_option(
         "-u", "--status",
-        help="Task job status file.",
-        action="store_true", default=False, dest="job_status")
-
-    parser.add_option(
-        "-c", "--custom",
```
Comment (arjclark) [inline bin/cylc-cat-log]: Looks like we've lost the -c functionality here? is that ok e.g. can we now no longer view job.stats files with this cmd?

File: bin/cylc-cat-log [hunk 3]
```diff
@@ -29,227 +29,276 @@ if remrun().execute():
...
+        GLOBAL_CFG.get_derived_host_item(
+            suite, "suite job log directory", host, owner),
+        point, task, submit_num, basename))
+
+
+def get_task_job_attrs(options, suite, point, task, submit_num):
+    """Return (user@host, command0) of a task job log.
+
+    user@host is set if task job is run remotely and for relevant log files.
+    command0 is set if task job if running on a batch system that requires a
```
Comment (arjclark) [inline bin/cylc-cat-log]: ...if task job *is* running...

File: bin/cylc-cat-log [hunk 4]
```diff
@@ -29,227 +29,276 @@ if remrun().execute():
...
+    """
+    if (options.filename in JOB_LOG_LOCAL_ALWAYS or
+             options.list_mode == LIST_MODE_LOCAL):
+        return (None, None)
+    suite_dao = CylcSuiteDAO(
+        os.path.join(
+            GLOBAL_CFG.get_derived_host_item(suite, "suite run directory"),
+            CylcSuiteDAO.DB_FILE_BASE_NAME),
+        is_public=True)
+    task_job_data = suite_dao.select_task_job(None, point, task, submit_num)
```
Comment (arjclark) [inline bin/cylc-cat-log]: are we happy to have another db connection being made? will this cope ok if the database is locked?

File: doc/cug.tex [hunk 5]
```diff
@@ -5224,6 +5224,27 @@ \subsection{Task stdout And stderr Logs}
...
 to here will be complete task job logs.
 
+Some job submission methods, such as \lstinline=pbs=, redirect a job's stdout
+and stderr streams to a separate cache area while the job is running. The
+contents are only copied to the normal locations when the job completes. This
+means that \lstinline=cylc log= or the gcylc GUI will be unable to find the
+job's stdout and stderr streams while the job is running. Some sites with these
+job submission methods are known to provide commands for viewing and/or
+tail-follow a job's stdout and stderr streams that are redirected to these
+cache areas. If this is the case at your site, you can make use of the facility
```
Comment (arjclark) [inline doc/cug.tex]: For clarity, how about:

If this is the case at your site, you can configure cylc to make use of the provided commands by adding ...

**Comment by arjclark** [comment]
The tex in the documentation doesn't seem to be correct, make test failing and running manually gives you:
```
...
! Missing $ inserted.
<inserted text> 
                $
l.686 The command template (with \%(job_
                                        id)s substitution) for testing the r...
!  ==> Fatal error occurred, no output PDF file produced!
Transcript written on cug-pdf.log.
make: *** [pdf/cug-pdf.pdf] Error 1
```

File: bin/cylc-cat-log (hunk 4 above)
Comment (matthewrmshin) [inline bin/cylc-cat-log]: This is not really another connection. The old logic already connects to the run time database. It used to select from the `task_states` table. It now selects from the `task_jobs` table.

Comment (arjclark) [inline bin/cylc-cat-log]: Cool. The blitzing of the old code meant I hadn't picked up on that.

**Comment by matthewrmshin** [comment]
Doc syntax fixed.

**Comment by arjclark** [comment]
Results as follows with *no* global.cfg additions

Error thrown when trying to view the jobfile via the gui:
```
gcylc has a problem.

Traceback (most recent call last):
  File "/net/home/h06/aclark/cylc.git/lib/cylc/gui/combo_logviewer.py", line 90, in switch_log
    if filename != self.filename:
UnboundLocalError: local variable 'filename' referenced before assignment
```

Also, for trying to view logs for general linux platforms running under background I can't view logs:
```
tail: cannot open `<host>:/<path>/cylc-run/<suite>/log/job/1/hello_orig_host/01/job.out' for reading: No such file or directory
tail: cannot watch parent directory of `<host>:<path>/cylc-run/<suite>/log/job/1/hello_orig_host/01/job.out': No such file or directory
```
perhaps the gui's not correctly using ssh?

Can be tested using mi-aa121 in my roses dir.

**Comment by matthewrmshin** [comment]
The reported GUI problem should now be fixed.

**Comment by matthewrmshin** [comment]
Branch re-based.

**Comment by arjclark** [comment]
*With* global.rc entries:

`cylc cat-log mi-aa121 hello_hpc.1 --host=host06 -f job.out` is fine but gui viewing of job.out and job.err just gives a blank screen for me.

**Comment by matthewrmshin** [comment]
> gui viewing of job.out and job.err just gives a blank screen for me.

There appears to be a local problem with `qcat -f` at the moment.

**Comment by arjclark** [comment]
Will await the qcat fix on Tuesday.

**Comment by arjclark** [comment]
All now working.

**Comment by hjoliver** [comment]
(Note that this one didn't go through two full reviews; I had just provided some requested "initial comment" above, before @matthewrmshin had finished with the coding.  However, I can't test it without access to PBS anyway, and it appears that a thorough review was done by @arjclark.)

**Comment by arjclark** [comment]
D'oh, sorry my mistake! I misread your "looks good" comment as being approval.

**Pull request merged**